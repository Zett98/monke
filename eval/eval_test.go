package eval

import (
	"bytes"
	"encoding/json"
	"monke/ast"
	"monke/lexer"
	"monke/parser"
	"os"
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
)

func parse(t *testing.T, s *string) ast.Node {
	lex := lexer.New(s)
	parser := parser.New(lex)
	prog := parser.ParseProgram()
	if err := parser.Errors(); len(err) != 0 {
		t.Logf("Unexpected error %v\n", err)
		t.FailNow()
	}
	return prog
}
func TestEval(t *testing.T) {
	type args struct {
		node ast.Node
	}
	tests := []struct {
		name string
		args string
	}{
		{
			"eval int",
			"5",
		},
		{
			"true",
			"true",
		},
		{
			"false",
			"false",
		},
		{
			"truthy `!true`",
			"!true",
		},
		{
			"truthy `!false`",
			"!false",
		},
		{
			"truthy `!5`",
			"!5",
		},
		{
			"truthy `!!true`",
			"!!true",
		},
		{
			"truthy `!!false`",
			"!!false",
		},
		{
			"truthy `!!5`",
			"!!5",
		},
		{
			"truthy `!!0`",
			"!!0",
		},
		{
			"prefix minus",
			"-5",
		},
		{
			"infix",
			"true == true",
		},
		{
			"infix",
			"false == false",
		},
		{
			"infix",
			"true == false",
		},
		{
			"infix",
			"true != false",
		},
		{
			"infix",
			"false != true",
		},
		{
			"infix",
			"(1 < 2) == true",
		},
		{
			"infix",
			"(1 < 2) == false",
		},
		{
			"infix",
			"(1 > 2) == true",
		},
		{
			"infix",
			"(1 > 2) == false",
		},
		{
			"cond",
			"if (1 < 2) { 10 }",
		},
		{
			"cond",
			"if (1 > 2) { 10 }",
		},
		{
			"cond",
			"if (1 > 2) { 10 } else { 20 }",
		},
		{
			"cond",
			"if (1 < 2) { 10 } else { 20 }",
		},
		{
			"return",
			"return 10;",
		},
		{
			"sample program",
			`
			if (10 > 1) {
				if (10 > 1) {
					return 10;
				}
				return 1;
			}
			`,
		},
		{
			"error",
			"5 + true",
		},
		{
			"environment",
			"let x = 5; return x + 5;",
		},
		{
			"environment",
			`let newAdder = fn(x) { fn(y) { x + y }; };
				let addTwo = newAdder(2);
				addTwo(2);`,
		},
		{
			"environment",
			"let f = fn(x) { let x = 10; x + 2; }; f(5);",
		},
		{
			"environment",
			"let x = 1; let f = fn(x) { x + 2; };  f(5);",
		},
	}
	for _, tt := range tests {
		env := NewEnvironment()
		t.Run(tt.name, func(t *testing.T) {
			parsed := parse(t, &tt.args)

			result := struct {
				Input string `json:"input"`
				Out   string `json:"out"`
			}{
				Input: tt.args,
				Out:   Eval(parsed, env).Inspect(),
			}
			buf := bytes.NewBufferString("")
			enc := json.NewEncoder(buf)
			enc.SetEscapeHTML(false)
			enc.SetIndent("", "  ")
			err := enc.Encode(result)
			if err != nil {
				t.Errorf("%v", err)
				t.FailNow()
			}
			// NB: object.Bool.false are omitted during printing
			// because `false` is a default value
			snaps.MatchSnapshot(t,
				buf.String())
		})
	}
}

func TestMain(m *testing.M) {
	v := m.Run()

	// After all tests have run `go-snaps` will sort snapshots
	snaps.Clean(m, snaps.CleanOpts{Sort: true})

	os.Exit(v)
}
