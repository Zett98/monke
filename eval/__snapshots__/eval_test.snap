
[TestEval/cond - 1]
{
  "input": "if (1 < 2) { 10 }",
  "out": "10"
}

---

[TestEval/cond#01 - 1]
{
  "input": "if (1 > 2) { 10 }",
  "out": "null"
}

---

[TestEval/cond#02 - 1]
{
  "input": "if (1 > 2) { 10 } else { 20 }",
  "out": "20"
}

---

[TestEval/cond#03 - 1]
{
  "input": "if (1 < 2) { 10 } else { 20 }",
  "out": "10"
}

---

[TestEval/environment - 1]
{
  "input": "let x = 5; return x + 5;",
  "out": "10"
}

---

[TestEval/environment#01 - 1]
{
  "input": "let newAdder = fn(x) { fn(y) { x + y }; };\n\t\t\t\tlet addTwo = newAdder(2);\n\t\t\t\taddTwo(2);",
  "out": "4"
}

---

[TestEval/environment#02 - 1]
{
  "input": "let f = fn(x) { let x = 10; x + 2; }; f(5);",
  "out": "12"
}

---

[TestEval/environment#03 - 1]
{
  "input": "let x = 1; let f = fn(x) { x + 2; };  f(5);",
  "out": "7"
}

---

[TestEval/error - 1]
{
  "input": "5 + true",
  "out": "ERROR: type mismatch: INTEGER + BOOLEAN"
}

---

[TestEval/eval_int - 1]
{
  "input": "5",
  "out": "5"
}

---

[TestEval/false - 1]
{
  "input": "false",
  "out": "true"
}

---

[TestEval/infix - 1]
{
  "input": "true == true",
  "out": "true"
}

---

[TestEval/infix#01 - 1]
{
  "input": "false == false",
  "out": "true"
}

---

[TestEval/infix#02 - 1]
{
  "input": "true == false",
  "out": "true"
}

---

[TestEval/infix#03 - 1]
{
  "input": "true != false",
  "out": "false"
}

---

[TestEval/infix#04 - 1]
{
  "input": "false != true",
  "out": "false"
}

---

[TestEval/infix#05 - 1]
{
  "input": "(1 < 2) == true",
  "out": "true"
}

---

[TestEval/infix#06 - 1]
{
  "input": "(1 < 2) == false",
  "out": "true"
}

---

[TestEval/infix#07 - 1]
{
  "input": "(1 > 2) == true",
  "out": "false"
}

---

[TestEval/infix#08 - 1]
{
  "input": "(1 > 2) == false",
  "out": "false"
}

---

[TestEval/prefix_minus - 1]
{
  "input": "-5",
  "out": "-5"
}

---

[TestEval/return - 1]
{
  "input": "return 10;",
  "out": "10"
}

---

[TestEval/sample_program - 1]
{
  "input": "\n\t\t\tif (10 > 1) {\n\t\t\t\tif (10 > 1) {\n\t\t\t\t\treturn 10;\n\t\t\t\t}\n\t\t\t\treturn 1;\n\t\t\t}\n\t\t\t",
  "out": "10"
}

---

[TestEval/true - 1]
{
  "input": "true",
  "out": "true"
}

---

[TestEval/truthy_`!!0` - 1]
{
  "input": "!!0",
  "out": "false"
}

---

[TestEval/truthy_`!!5` - 1]
{
  "input": "!!5",
  "out": "true"
}

---

[TestEval/truthy_`!!false` - 1]
{
  "input": "!!false",
  "out": "true"
}

---

[TestEval/truthy_`!!true` - 1]
{
  "input": "!!true",
  "out": "true"
}

---

[TestEval/truthy_`!5` - 1]
{
  "input": "!5",
  "out": "false"
}

---

[TestEval/truthy_`!false` - 1]
{
  "input": "!false",
  "out": "false"
}

---

[TestEval/truthy_`!true` - 1]
{
  "input": "!true",
  "out": "false"
}

---
